cpp/src/io/utilities/hostdevice_vector.hpp

```diff
@@ -81,17 +81,26 @@ class hostdevice_vector {
   [[nodiscard]] size_t size() const noexcept { return num_elements; }
   [[nodiscard]] size_t memory_size() const noexcept { return sizeof(T) * num_elements; }
 
-  T& operator[](size_t i) const { return h_data[i]; }
-  T* host_ptr(size_t offset = 0) const { return h_data + offset; }
+  T& operator[](size_t i) { return h_data[i]; }
+  T const& operator[](size_t i) const { return h_data[i]; }
+
+  T* host_ptr(size_t offset = 0) { return h_data + offset; }
+  T const* host_ptr(size_t offset = 0) const { return h_data + offset; }
+
   T* begin() { return h_data; }
+  T const* begin() const { return h_data; }
+
   T* end() { return h_data + num_elements; }
-  T* d_begin() { return static_cast<T*>(d_data.data()); }
-  T* d_end() { return static_cast<T*>(d_data.data()) + num_elements; }
-  T* device_ptr(size_t offset = 0) { return reinterpret_cast<T*>(d_data.data()) + offset; }
-  T const* device_ptr(size_t offset = 0) const
-  {
-    return reinterpret_cast<T const*>(d_data.data()) + offset;
-  }
+  T const* end() const { return h_data + num_elements; }
+
+  auto d_begin() { return static_cast<T*>(d_data.data()); }
```

Comment:
For clarity and consistency with the host accessors, should we use `T*`, `T const*`, etc. instead of `auto`?

Reply:
Yeah, I justified it with the cast in each function (strictly speaking, naming the type twice breaks the DRY rule).
The real reason is that the last function breaks into multiple lines if I specify the return type :D

I don't have any good reasons to prefer one of the options here, let me know if you want to remove `auto`s here.

Reply:
I approved the PR because I was fine with any outcome here. Just leave it as-is.